src/win/fs.c

```diff
@@ -2627,8 +2627,56 @@ static void fs__statfs(uv_fs_t* req) {
                              &bytes_per_sector,
                              &free_clusters,
                              &total_clusters)) {
-    SET_REQ_WIN32_ERROR(req, GetLastError());
-    return;
+    DWORD err;
+
+    err = GetLastError();
+    if (err == ERROR_DIRECTORY) {
+      WCHAR* pathw;
+      WCHAR* fpart;
+      size_t len;
+      DWORD ret;
+
+      len = MAX_PATH + 1;
+      pathw = uv__malloc(len * sizeof(*pathw));
+retry:
+      ret = GetFullPathNameW(req->file.pathw,
+                             len,
+                             pathw,
+                             &fpart);
+      if (ret == 0) {
+        uv__free(pathw);
+        SET_REQ_WIN32_ERROR(req, err);
+        return;
+      } else if (ret > len) {
+        WCHAR* saved_pathw;
+
+        len = ret;
+        saved_pathw = pathw;
+        pathw = uv__realloc(pathw, len * sizeof(*pathw));
+        if (pathw == NULL) {
+          uv__free(saved_pathw);
+          SET_REQ_WIN32_ERROR(req, err);
+          return;
+        }
+        goto retry;
+      }
+      if (fpart != 0)
+        *fpart = L'\0';
+
+      if (0 == GetDiskFreeSpaceW(pathw,
```

Comment:
Instead of having a second `GetDiskFreeSpaceW()` call, would it be possible to have `pathw` initially point to `req->file.pathw`, and have the retry logic jump back to the first `GetDiskFreeSpaceW()` call? You could check `pathw == req->file.pathw` to make sure you don't try to reallocate or free `req->file.pathw`.